optimism/src/main.rs

```diff
@@ -40,12 +40,113 @@ pub fn main() -> ExitCode {
+
+        let mut srs = poly_commitment::pairing_proof::PairingSRS::create(x, domain_size);
+        srs.full_srs.add_lagrange_basis(domain.d1);
+        srs
+    };
+
+    let mut env = witness::Env::<ark_bn254::Fr>::create(cannon::PAGE_SIZE as usize, state, po);
+
+    let mut accumulator = proof::ProofInputs::<
+        ark_ec::short_weierstrass_jacobian::GroupAffine<ark_bn254::g1::Parameters>,
+    >::default();
+
+    let new_chunk = || proof::WitnessColumns {
+        scratch: std::array::from_fn(|_| Vec::with_capacity(domain_size)),
+        instruction_counter: Vec::with_capacity(domain_size),
+        error: Vec::with_capacity(domain_size),
+    };
+
+    let mut current_chunk = new_chunk();
+
+    use mina_poseidon::{
+        constants::PlonkSpongeConstantsKimchi,
+        sponge::{DefaultFqSponge, DefaultFrSponge},
+    };
+    type Fp = ark_bn254::Fr;
+    type SpongeParams = PlonkSpongeConstantsKimchi;
+    type BaseSponge = DefaultFqSponge<ark_bn254::g1::Parameters, SpongeParams>;
+    type ScalarSponge = DefaultFrSponge<Fp, SpongeParams>;
 
     while !env.halt {
         env.step(&configuration, &meta, &start);
+        for (scratch, scratch_chunk) in env
+            .scratch_state
+            .iter()
+            .zip(current_chunk.scratch.iter_mut())
+        {
+            scratch_chunk.push(*scratch);
+        }
+        current_chunk
+            .instruction_counter
+            .push(ark_bn254::Fr::from(env.instruction_counter));
+        // TODO
+        use ark_ff::UniformRand;
+        current_chunk
+            .error
+            .push(ark_bn254::Fr::rand(&mut rand::rngs::OsRng));
+        if current_chunk.instruction_counter.len() == 1 << 15 {
+            proof::fold::<
+                _,
+                poly_commitment::pairing_proof::PairingProof<ark_ec::bn::Bn<ark_bn254::Parameters>>,
+                BaseSponge,
+                ScalarSponge,
+            >(domain, &srs, &mut accumulator, current_chunk);
+            current_chunk = new_chunk();
+        }
+    }
+    if !current_chunk.instruction_counter.is_empty() {
+        use ark_ff::Zero;
+        let remaining = domain_size - current_chunk.instruction_counter.len();
+        for scratch in current_chunk.scratch.iter_mut() {
+            scratch.extend((0..remaining).map(|_| ark_bn254::Fr::zero()));
+        }
+        current_chunk
+            .instruction_counter
+            .extend((0..remaining).map(|_| ark_bn254::Fr::zero()));
+        current_chunk
+            .error
+            .extend((0..remaining).map(|_| ark_bn254::Fr::zero()));
+        proof::fold::<
+            _,
+            poly_commitment::pairing_proof::PairingProof<ark_ec::bn::Bn<ark_bn254::Parameters>>,
+            BaseSponge,
+            ScalarSponge,
+        >(domain, &srs, &mut accumulator, current_chunk);
+    }
+
+    {
+        let proof = proof::prove::<
+            _,
+            poly_commitment::pairing_proof::PairingProof<ark_ec::bn::Bn<ark_bn254::Parameters>>,
```

Comment:
I would personally try to import these up-front so the code does not look so verbose. Otherwise it is a bit hard to parse what is actually being called.